### DISCUSSION TITLE
[Berry] Best way to detect if wifi is connected?

### DISCUSSION THREAD

**Discussion by Slyke** [arendst/Tasmota#15887] [Q&A] [answered]
Is the best way to detect if the wifi is connected or not to do something like:

```
var wifiDetails = tasmota.wifi()
if wifiDetails.contains('quality') && wifiDetails['quality'] > 0
  # Is connected
end
```

There seems to be no way to know if `mqtt.publish('topic', 'payload', false)` was successful or not as it doesn't seem to return anything, so thinking to check if wifi is connected as the next best way.

**Comment by sfromis** [comment, marked as answer]
What I see is that the map from `tasmota.wifi()` is empty when not connected, both in the case of doing a `Wifi 0` command or just yanking the power of the AP. Unsurprisingly, there is a delay from yanking to Tasmota giving up on Wifi, and any attempted publish in that time frame will not arrive.

One viable strategy to see what Tasmota thinks is `tasmota.eth().find('ip') != nil`. Looking at the "quality" number when connected gives no strong indicator of reliability, but of course very weak signals increases the risk of data being lost.

As QoS > 0 is not supported, there is no 100% way of detecting if a publish was successful or not. If you need such an indicator, have the subscribing client publish a confirmation message about success.

One easy way of creating a test setup is to configure some Tasmota device as a range extender, giving you a small AP outside the potential "inconvenience" of shutting down the house Wifi for testing.

**Reply by Slyke** [reply]
Thank you! That works and makes sense.
